### PR TITLE
Modify ApplicationSampler to user ApplicationStatus

### PR DESCRIPTION
**Pull request by cmcantalupo**

### CODE REVIEW

File: src/ApplicationSampler.hpp [hunk 1]
```diff
@@ -85,14 +87,8 @@ namespace geopm
...
             virtual std::map<uint64_t, std::string> get_name_map(uint64_t name_key) const = 0;
-            /// @brief Sample the current hint for every cpu.
-            /// @return Vector over Linux logical CPU of the GEOPM
-            ///         region hint currently being executed.
-            virtual std::vector<uint64_t> per_cpu_hint(void) const = 0;
-            /// @brief Sample the current progress for every cpu.
-            /// @return Vector over Linux logical CPU of the region
-            ///         progress.
-            virtual std::vector<double> per_cpu_progress(void) const = 0;
+            virtual double cpu_hint_time(int cpu_idx, uint64_t hint) const = 0;
```
Comment (cmcantalupo) [inline src/ApplicationSampler.hpp:118]: Doxygen missing

File: src/ApplicationSampler.hpp [hunk 2]
```diff
@@ -67,8 +67,10 @@ namespace geopm
             virtual void time_zero(const geopm_time_s &start_time) = 0;
             /// @brief Update the record buffer by clearing out old
             ///        records and providing a new cache for
-            ///        subsequent calls to the get_records() method.
-            virtual void update_records(void) = 0;
+            ///        subsequent calls to the get_records()
+            ///        method. Also update cache of application
+            ///        status for use by hint and progress APIs.
+            virtual void update(const geopm_time_s &curr_time) = 0;
```
Comment (cmcantalupo) [inline src/ApplicationSampler.hpp:73]: document curr_time

**Review by cmcantalupo — COMMENTED** [review]
Needs some dox and some unit tests

File: src/ApplicationSampler.cpp [hunk 3]
```diff
@@ -108,20 +111,41 @@ namespace geopm
...
+                                std::map<int, m_process_s> {},
                                 environment().do_record_filter(),
                                 environment().record_filter())
     {
 
     }
 
-    ApplicationSamplerImp::ApplicationSamplerImp(const std::map<int, m_process_s> &process_map,
+const std::map<uint64_t, double> ApplicationSamplerImp::m_hint_time_init = {
+    {GEOPM_REGION_HINT_UNKNOWN, 0.0},
```
Comment (dianarg) [inline src/ApplicationSampler.cpp]: indentation

File: src/Controller.cpp [hunk 4]
```diff
@@ -354,8 +354,10 @@ namespace geopm
         m_reporter->init();
         setup_trace();
         m_application_io->controller_ready();
-
         m_application_io->update(m_comm);
+        geopm_time_s curr_time;
+        geopm_time(&curr_time);
+        m_application_sampler.update(curr_time);
```
Comment (dianarg) [inline src/Controller.cpp:360]: it would be nice to have a function returning the current time

File: src/ApplicationSampler.hpp (hunk 2 above)
Comment (dianarg) [reply src/ApplicationSampler.hpp:73]: this is exposed for testing?

File: src/ApplicationSampler.cpp [hunk 5]
```diff
@@ -131,8 +155,18 @@ namespace geopm
...
     }
 
-    void ApplicationSamplerImp::update_records(void)
+    void ApplicationSamplerImp::update(const geopm_time_s &curr_time)
     {
+        m_status->update_cache();
+        if (!m_is_first_update) {
+            double time_delta = geopm_time_diff(&m_update_time, &curr_time);
+            for (int cpu_idx = 0; cpu_idx != m_num_cpu; ++cpu_idx) {
+                uint64_t hint = m_status->get_hint(cpu_idx);
```
Comment (dianarg) [inline src/ApplicationSampler.cpp:164]: status needs to guarantee hint is in range; no multiple hints

File: src/ApplicationSampler.hpp (hunk 2 above)
Comment (cmcantalupo) [reply src/ApplicationSampler.hpp:73]: Yes...might be nice to have a Clock singleton that we could mock for time values.

File: test/ApplicationSamplerTest.cpp [hunk 6]
```diff
@@ -294,9 +302,9 @@ TEST_F(ApplicationSamplerTest, string_conversion)
 
 TEST_F(ApplicationSamplerTest, process_mapping)
 {
-    std::vector<int> expected { 4, 5, 77, 32 };
-    EXPECT_CALL(*m_mock_profile_sampler, cpu_rank())
-        .WillOnce(Return(expected));
+    std::vector<int> expected {42, 42};
+    EXPECT_CALL(*m_mock_status, get_process(_))
+        .WillRepeatedly(Return(42));
```
Comment (dianarg) [inline test/ApplicationSamplerTest.cpp]: if you want to test different values here, you can chain together multiple ` .WillOnce(Return(..))`

File: test/ApplicationSamplerTest.cpp [hunk 7]
```diff
@@ -359,3 +368,86 @@ TEST_F(ApplicationSamplerTest, short_regions)
...
+                            SetArgReferee<1>(empty_short_region_buffer)));
+        EXPECT_CALL(*m_record_log_1, dump(_, _))
+            .WillOnce(DoAll(SetArgReferee<0>(empty_message_buffer),
+                            SetArgReferee<1>(empty_short_region_buffer)));
+        EXPECT_CALL(*m_mock_status, update_cache());
+        EXPECT_CALL(*m_mock_status, get_hint(_))
+            .WillRepeatedly(Return(GEOPM_REGION_HINT_NETWORK));
+        m_app_sampler->update({{2, 0}});
+    }
+    double compute_time = m_app_sampler->cpu_hint_time(0, GEOPM_REGION_HINT_COMPUTE);
```
Comment (dianarg) [inline test/ApplicationSamplerTest.cpp]: intermediate variables here are not really making the lines shorter

**Review by dianarg — APPROVED** [review]

**Pull request merged**